### PR TITLE
SALTO-3743: Hide name for custom applications in Okta 

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Hide name for custom applications in Okta 

---

When adding a new custom `Application` instance in Okta, `name` field must be blank, but when `Application` is created the service generates a name for it that we need to use when modifying the `Application`.
`onFetch` we'll create a hidden value for custom app name (`customName`) & delete the original `name` field. When adding a new custom application the generated name will be assigned to `customName` field. 

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/okta-adapter/src/filters/app_deployment.ts [hunk 1]
```diff
@@ -121,14 +189,28 @@ const filterCreator: FilterCreator = ({ client, config }) => ({
...
       deployResult,
     }
   },
+  onDeploy: async (changes: Change<InstanceElement>[]) => {
+    changes
+      .map(getChangeData)
+      .filter(isInstanceElement)
+      .filter(instance => instance.elemID.typeName === APPLICATION_TYPE_NAME)
+      .forEach(
+        async instance => {
```
Comment (edenhassid) [inline packages/okta-adapter/src/filters/app_deployment.ts:206]: why async?

File: packages/okta-adapter/test/filters/app_deployment.test.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    client = cli
+    filter = appDeploymentFilter(
+      getFilterParams({ client, elementsSource: buildElementsSourceFromElements([orgSettingInstance]) })
+    ) as typeof filter
+  })
+
+  describe('fetch', () => {
+    it('should create customName field for custom applications', async () => {
+      const elements = [appType, orgSettingType, orgSettingInstance, customSamlAppInstance, customSwaInstance]
+      await filter.onFetch?.(elements)
```
Comment (edenhassid) [inline packages/okta-adapter/test/filters/app_deployment.test.ts:78]: why the `?`

**Review by edenhassid — COMMENTED** [review]
looks good! added a few comments :)

File: packages/okta-adapter/src/filters/app_deployment.ts [hunk 3]
```diff
@@ -48,59 +55,87 @@ const APP_ASSIGNMENT_FIELDS: Record<string, configUtils.DeploymentRequestsByActi
...
   client: OktaClient,
   config: OktaConfig,
+  elementsSource: ReadOnlyElementsSource,
 ): Promise<void> => {
-  const getAdditionalFieldsToIgnore = (instance: InstanceElement): string[] =>
-    (instance.value.signOnMode === AUTO_LOGIN_APP ? ['name'] : [])
-
   const fieldsToIgnore = [
     ...Object.keys(APP_ASSIGNMENT_FIELDS),
     // TODO SALTO-2690: remove this once completed
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/filters/app_deployment.ts:122]: I'll remove this in the next PR, u can ignore for now (:

File: packages/okta-adapter/src/filters/app_deployment.ts [hunk 4]
```diff
@@ -48,59 +55,87 @@ const APP_ASSIGNMENT_FIELDS: Record<string, configUtils.DeploymentRequestsByActi
...
     // Application is created with status 'ACTIVE' by default, unless we provide activate='false' as a query param
-    isAdditionChange(change) && getChangeData(change).value.status === INACTIVE_STATUS ? { activate: 'false' } : undefined
+    isAdditionChange(change) && getChangeData(change).value?.status === INACTIVE_STATUS ? { activate: 'false' } : undefined
   )
 
   if (isAdditionOrModificationChange(change)) {
-    if (isAdditionChange(change)) {
-      assignCreatedFieldsToApp(change, response)
+    if (isAdditionChange(change) && isAppResponse(response)) {
+      const subdomain = await getSubdomainFromElementsSource(elementsSource)
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/filters/app_deployment.ts:137]: good point, thanks!

File: packages/okta-adapter/src/filters/app_deployment.ts [hunk 5]
```diff
@@ -48,59 +55,87 @@ const APP_ASSIGNMENT_FIELDS: Record<string, configUtils.DeploymentRequestsByActi
...
+  name: Joi.string().required(),
+  signOnMode: Joi.string().required(),
+}).unknown(true)
+
+export const isAppResponse = createSchemeGuard<Application>(EXPECTED_APP_SCHEMA, 'Received an invalid application response')
+
+const isCustomApp = (value: Values, subdomain: string): boolean => (
+  [AUTO_LOGIN_APP, SAML_2_0_APP].includes(value.signOnMode)
+  && value.name !== undefined
+  && _.startsWith(value.name, subdomain)
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/filters/app_deployment.ts:75]: Thanks!

File: packages/okta-adapter/src/config.ts [hunk 6]
```diff
@@ -338,6 +339,7 @@ const DEFAULT_TYPE_CUSTOMIZATIONS: OktaApiConfig['types'] = {
       idFields: ['label'],
       serviceIdField: 'id',
       fieldsToHide: [
+        { fieldName: 'customName' },
```
Comment (edenhassid) [inline packages/okta-adapter/src/config.ts:342]: nit: you can add the constant also here

File: packages/okta-adapter/test/filters/app_deployment.test.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      const saml = elements.filter(isInstanceElement).find(e => e.elemID.name === 'custom saml app')
+      expect(saml?.value.name).toBeUndefined()
+      expect(saml?.value.customName).toEqual('oktaSubdomain_saml_link')
+      const swa = elements.filter(isInstanceElement).find(e => e.elemID.name === 'custom swa app')
+      expect(swa?.value.name).toBeUndefined()
+      expect(swa?.value.customName).toEqual('oktaSubdomain_swa_link')
+    })
+    it('should not create customName field for non custom applications and remain name field', async () => {
+      const elements = [appType, orgSettingType, orgSettingInstance, appInstance]
+      await filter.onFetch?.(elements)
```
Comment (edenhassid) [inline packages/okta-adapter/test/filters/app_deployment.test.ts:88]: nit: I think you can remove the `?`

File: packages/okta-adapter/test/filters/app_deployment.test.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+      const instances = changes.map(getChangeData).filter(isInstanceElement)
+      const app = instances.find(i => i.elemID.name === 'regular app')
+      const customApp = instances.find(i => i.elemID.name === 'custom saml app')
+      expect(app?.value).toEqual({ name: 'salesforce', signOnMode: 'SAML_2_0' })
+      expect(customApp?.value).toEqual({ signOnMode: 'SAML_2_0' })
+    })
+  })
+
+  describe('deploy', () => {
+    it('should assign customName field to custom app on addition change', async () => {
```
Comment (edenhassid) [inline packages/okta-adapter/test/filters/app_deployment.test.ts:125]: maybe add a test where the deploy fails? the response is not valid or something like that

**Review by edenhassid — APPROVED** [review]
looks great! 🔥 added a few nits and one unit test suggestion :)

File: packages/okta-adapter/test/filters/app_deployment.test.ts (hunk 8 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/test/filters/app_deployment.test.ts:125]: In this PR I added just relevant tests for the customApp field
In the next PR which implements the whole app flow I'll add all relevant tests :)

**Pull request merged**